File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
#   preorder = [1,2,3,4], inorder = [2,1,3,4]

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        map = {val : idx for idx, val in enumerate(inorder)}
        self.i = 0

        def dfs(l , r):
            if l > r:
                return None
            
            rootValue = preorder[self.i]
            self.i += 1
            root = TreeNode(rootValue)
            mid = map.get(rootValue)
            root.left = dfs(l , mid -1)
            root.right = dfs( mid + 1, r)
            return root
            
        return dfs(0, len(inorder) - 1)
        

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public List<Integer> rightSideView(TreeNode root) {

        List<Integer> answer = new ArrayList<>();
        if(root == null)
            return answer;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);

        while(!queue.isEmpty())
        {  
            int size = queue.size();
            TreeNode rightMost = null;
            for(int i = 0; i < size; i++)
            {
                TreeNode node = queue.poll();
                if(node.left != null)
                    queue.offer(node.left);
                if(node.right != null)
                    queue.offer(node.right);

                rightMost = node; // the item added last in queue needs to come so keep updating
   
            }
            answer.add(rightMost.val);

        }

        return answer;

        
    }
}
